test/invariants/fuzz/Protocol.t.sol

```diff
@@ -0,0 +1,568 @@
+  /// @custom:property there can't be any amount out for a 0 amount in
+  /// @custom:property-id 19
+  /// @custom:property a swap can only happen when the pool is finalized
+  function fuzz_swapExactOut(
+    uint256 _maxAmountIn,
+    uint256 _amountOut,
+    uint256 _tokenIn,
+    uint256 _tokenOut
+  ) public AgentOrDeployer {
+    // Precondition
+    require(pool.isFinalized());
+
+    _tokenIn = clamp(_tokenIn, 0, tokens.length - 1);
+    _tokenOut = clamp(_tokenOut, 0, tokens.length - 1);
+    _amountOut = clamp(_amountOut, 1 ether, 10 ether);
+    _maxAmountIn = clamp(_maxAmountIn, 1 ether, 10 ether);
+
+    tokens[_tokenIn].mint(currentCaller, _maxAmountIn);
+
+    hevm.prank(currentCaller);
+    tokens[_tokenIn].approve(address(pool), type(uint256).max); // approval isn't limiting
+
+    uint256 _balanceInBefore = tokens[_tokenIn].balanceOf(currentCaller);
+    uint256 _balanceOutBefore = tokens[_tokenOut].balanceOf(currentCaller);
+
+    uint256 _inComputed = bmath.calcInGivenOut(
+      tokens[_tokenIn].balanceOf(address(pool)),
+      pool.getDenormalizedWeight(address(tokens[_tokenIn])),
+      tokens[_tokenOut].balanceOf(address(pool)),
+      pool.getDenormalizedWeight(address(tokens[_tokenOut])),
+      _amountOut,
+      bconst.MIN_FEE()
+    );
+
+    hevm.prank(currentCaller);
+
+    // Action
+    try pool.swapExactAmountOut(
+      address(tokens[_tokenIn]), _maxAmountIn, address(tokens[_tokenOut]), _amountOut, type(uint256).max
+    ) {
+      // Postcondition
+      uint256 _balanceInAfter = tokens[_tokenIn].balanceOf(currentCaller);
+      uint256 _balanceOutAfter = tokens[_tokenOut].balanceOf(currentCaller);
+
+      // Take into account previous direct transfers (only way to get free token)
+      uint256 _tokenOutInExcess = ghost_amountDirectlyTransfered[tokens[_tokenOut]] > _amountOut
+        ? _amountOut
+        : ghost_amountDirectlyTransfered[tokens[_tokenOut]];
+      ghost_amountDirectlyTransfered[tokens[_tokenOut]] -= _tokenOutInExcess;
+
+      // 5
+      assert(_balanceInBefore - _balanceInAfter <= _maxAmountIn);
+
+      // 14
+      if (_tokenIn != _tokenOut) assert(_balanceOutAfter - _balanceOutBefore == _amountOut);
+      else assert(_balanceOutAfter == _balanceOutBefore - _inComputed + _amountOut);
+
+      // 15
+      if (_balanceInBefore == _balanceInAfter) assert(_balanceOutBefore + _tokenOutInExcess == _balanceOutAfter);
+
+      // 19
+      assert(pool.isFinalized());
+    } catch {
+      uint256 _spotBefore = bmath.calcSpotPrice(
+        tokens[_tokenIn].balanceOf(address(pool)),
+        pool.getDenormalizedWeight(address(tokens[_tokenIn])),
+        tokens[_tokenOut].balanceOf(address(pool)),
+        pool.getDenormalizedWeight(address(tokens[_tokenOut])),
+        bconst.MIN_FEE()
+      );
+
+      uint256 _outRatio = bnum.bmul_exposed(tokens[_tokenOut].balanceOf(address(pool)), bconst.MAX_OUT_RATIO());
+
+      assert(
+        _inComputed > _maxAmountIn // 5
```

Comment:
this assert seems to always short-circuit on the first condition

Reply:
why do you think so? Test fails if I remove any of the 3 conditions for me

Reply:
then it's probably related to me running only 50k calls